### PR TITLE
Allow setting CC, CXX, and other variables to indicate compiler path

### PR DESCRIPTION
**Pull request by mppf**
This PR implements changes to how users can indicate a particular path to
a particular compiler. It is sort of a follow-up to PR #17800 which added
`CHPL_LLVM_PREFIX` for selecting an LLVM installation.

This PR makes the following variables settable by users (as environment
variables or in a chplconfig):
 * for setting all the compilers (host and target) at once using
   relatively well known environment variable names:
    * CC, CXX
 * for setting host and target compilers separately:
    * `CHPL_HOST_CC`, `CHPL_HOST_CXX`
    * `CHPL_TARGET_CC`, `CHPL_TARGET_CXX`
 * for indicating which system LLVM installation should be used / the
   path to it by providing the path to the llvm-config command:
    * `CHPL_LLVM_CONFIG`

Additionally it does some clean-ups for LLVM Makefiles.

For printchplenv, the new variables are in the compiler or runtime
categories but aren't printed by default (so `printchplenv --all` will
print them). This PR adjusts `printchplenv.py` to include an idea that
some variables shouldn't be in `--path` output since it doesn't make
sense to store the path to CC say in the compiler/runtime build path.

PR #17931 contains the documentation changes to go along with these changes.

Reviewed by @daviditen - thanks!

- [x] full local testing

Tricky Pieces:
 * are we happy with the names `CHPL_HOST_CC` etc ? These names are
   chosen to both try to be intuitive / map from `CC` / `CXX` but also
   because autotools projects might have HOSTCC that has a different
   meaning. (At least autotools arguably a different meaning for host
   than what we use). An interesting alternative is `CHPL_CC_HOST` but
   `CHPL_HOST_CC` fits better with `CHPL_TARGET_COMPILER`.
 * should `printchplenv --all` show `CHPL_TARGET_CC` etc and
   `CHPL_LLVM_CONFIG` ? In the PR it does, but these don't show up with
   `printchplenv` without `--all`.
 * should `printchplenv` show `CC` or `CXX` being set in some way? In the
   PR it does not, other than their values showing up in `CHPL_TARGET_CC`
   etc.
 * If one of `CC` and `CXX` is set but not both, should we warn? The PR
   does warn for this unless the CC value is the default for that
   compiler family (e.g. CC=gcc means we know CXX=g++). I think the
   warning is important because the variable `CC` is more well known and
   we have lots of code that might use a mix of C and C++ (e.g. in the
   compiler or in the runtime when including C++ libraries like RE2)
 * If e.g. `CHPL_TARGET_CC` is set but `CHPL_TARGET_COMPILER` is not,
   should we infer `CHPL_TARGET_COMPILER` the way we would for `CC`?
   Currently the PR does not do this and additionally it does not warn in
   that case.

 Future Work:
  * adding testing for setting `CHPL_HOST_CC` / `CC` / `CHPL_LLVM_CONFIG` etc.

### CODE REVIEW

**Comment by mppf** [comment]
I have to fix these errors:

```
[Error matching compiler output for interop/python/errorMessages/badFormalIntent-array]
[Error matching compiler output for interop/python/errorMessages/badFormalIntent-cstring]
[Error matching compiler output for interop/python/errorMessages/badFormalIntent]
[Error matching compiler output for interop/python/errorMessages/badReturnIntent]
[Error matching debug info for llvmDebug_test]
[Error matching compiler output for runtime/sungeun/chpl-env-gen]
```

**Review by stonea — COMMENTED** [review]
Hi Michael,

Just a few questions I have about this PR:

* I assume host refers to the compiler being used to compile Chapel (the compiler) itself and target refers to the compiler used to compile emitted .c\llvmir from Chapel. What is `CHPL_MAKE_COMPILER` then?

* Is `CHPL_LLVM_CONFIG` meant to only set the path for LLVM is or is intended to set some other parameters? If just the path I think I'd name it something like `CHPL_LLVM_PATH`

* Do we have tests that cover all these different configuration options? I guess a lot of it gets covered with our Jenkins configurations. Does anything test  `CHPL_LLVM_CONFIG`? I imagine getting a test set up for that might be a bit of a pain, but it also seems like the sort of thing that could regress and no one would catch it since most of us are probably either using the bundled LLVM or the default system LLVM.

---

Just a brainstorming idea:

It seems like we need to have a lot of logic spread throughout the makefiles to validate \ set defaults to these environmental variables. I think if I added a new variable it may not be obvious to me where to configure it, and having all this logic embedded inside makefiles may make it harder to move to a different make system (cmake or something) later on if we decide to do that. Would it be worth factoring out all this logic into a single Python script? The make system could then just invoke this script and if it returns successfully assume everything has been configured sanely and proceed.

**Comment by mppf** [comment]
> * I assume host refers to the compiler being used to compile Chapel (the compiler) itself and target refers to the compiler used to compile emitted .c\llvmir from Chapel. What is `CHPL_MAKE_COMPILER` then?

Yes, that's right. `CHPL_MAKE_COMPILER` is used inside the Makefiles and is the compiler appropriate for the current job (which is either compiling the compiler, the runtime, or a generated C program).

Also, a nit, in the case of the LLVM backend, we generate LLVM IR and then emit an object file and use `CHPL_TARGET_COMPILER` to link.

> 
> * Is `CHPL_LLVM_CONFIG` meant to only set the path for LLVM is or is intended to set some other parameters? If just the path I think I'd name it something like `CHPL_LLVM_PATH`

It sets the path to the `llvm-config` command in a corresponding way to `CC` / `CHPL_CC_HOST`. I am not excited about changing it to `CHPL_LLVM_PATH` both because it's after the deadline I asked for comments on the design and also because it would cause me to expect that setting it to say `/usr/local/opt/llvm-11` would work. But that is not what the variable means - one needs to provide `/usr/local/opt/llvm-11/bin/llvm-config` in that event. Note that the program is not always called literally `llvm-config` and at a different path. On some systems it might be called `llvm-config-11` for example.

This is described in the accompanying documentation PR -- https://github.com/chapel-lang/chapel/pull/17931/files#diff-b71d5b3486fed20a9661b02deca46e394853b16cabd6a320e987ea14fbbc067bR711 

> * Do we have tests that cover all these different configuration options? I guess a lot of it gets covered with our Jenkins configurations. Does anything test  `CHPL_LLVM_CONFIG`? I imagine getting a test set up for that might be a bit of a pain, but it also seems like the sort of thing that could regress and no one would catch it since most of us are probably either using the bundled LLVM or the default system LLVM.

`CHPL_LLVM_CONFIG` is set by the llvm-by-default detection of the system LLVM and we have that working in a few different ways on different systems. As a result I would expect we would notice if there is something severely wrong with it.

About the testing, the same could be said for `CC` / `CXX` / `CHPL_HOST_CC` etc. I agree it would be nice to add tests for these but I'm not quite sure how to do. I am open to feedback but I would not consider it a requirement for this PR.

> Just a brainstorming idea:
> 
> It seems like we need to have a lot of logic spread throughout the makefiles to validate \ set defaults to these environmental variables. I think if I added a new variable it may not be obvious to me where to configure it, and having all this logic embedded inside makefiles may make it harder to move to a different make system (cmake or something) later on if we decide to do that. Would it be worth factoring out all this logic into a single Python script? The make system could then just invoke this script and if it returns successfully assume everything has been configured sanely and proceed. 

Yes I am of the opinion that we are better off if we can gradually move some of this logic to util/chplenv Python code. This PR does that for the find-llvm-config.sh script. I think it'd be nice to move the compiler flags we set for different compilers to the Python code as well but I am uncertain if that will have broad support.

**Review by daviditen — APPROVED** [review]
Looks good, thanks!

**Comment by stonea** [comment]
>> Is CHPL_LLVM_CONFIG meant to only set the path for LLVM is or is intended to set some other parameters? If just the path I think I'd name it something like CHPL_LLVM_PATH

> It sets the path to the llvm-config command in a corresponding way to CC / CHPL_CC_HOST. I am not excited about changing it to CHPL_LLVM_PATH both because it's after the deadline I asked for comments on the design and also because it would cause me to expect that setting it to say /usr/local/opt/llvm-11 would work. But that is not what the variable means - one needs to provide /usr/local/opt/llvm-11/bin/llvm-config in that event. Note that the program is not always called literally llvm-config and at a different path. On some systems it might be called llvm-config-11 for example.

I see, thanks for clarifying.

> About the testing, the same could be said for CC / CXX / CHPL_HOST_CC etc. I agree it would be nice to add tests for these but I'm not quite sure how to do. I am open to feedback but I would not consider it a requirement for this PR.

For sure, if we wanted to add tests that would be its own project. I just get nervous when code is submitted without any sort of regression test. But this is the sort of thing where the time investment to make a test may not outweigh the risk of regression and maybe our existing Jenkins configurations already covers a lot of what that kind of regression test would be used for.
If we wanted to go that route. I imagine we’d need to do the following:

* To test what the host compiler is we could have some kind of hidden flag you pass to chpl that would have it print it out. Just use compiler specific macros like `__clang__` or `__GNUC__` to discover what compiler is being used. We’d then have to configure Jenkins to build chpl in a couple different compilers (I think maybe we already do?) and run this test.
* To test the target compiler have a flag passed to chpl that prints out the commands used to compile (I think we might already have this).

File: util/chplenv/chpl_compiler.py [hunk 1]
```diff
@@ -39,6 +39,60 @@ def get_prgenv_compiler():
...
+        cxx_compiler = get_compiler_from_command(cxx_val)
+
+    if cc_val and cxx_val:
+        if cc_compiler == cxx_compiler:
+            compiler_val = cc_compiler
+        else:
+            error("Conflicting compiler families for CC and CXX settings\n"
+                  "  {0} -> {1}\n"
+                  "  {2} -> {3}".format(cc_val, cc_val_command,
+                                        cxx_val, cxx_val_command))
```
Comment (ronawho) [inline util/chplenv/chpl_compiler.py:69]: Looks like `cc_val_command` and `cxx_val_command` are never defined so this result in a NameError exception and not our error message

Comment (mppf) [reply util/chplenv/chpl_compiler.py:69]: PR #18052 will fix

**Comment by bradcray** [comment]
Hi @mppf — It seems that something about this PR broke my fix for LLVM@11 on Mac Mojave (PR #17851).  It took me from this afternoon until now to bisect it down to this PR, and I haven't looked into the PR itself to try to understand how it could've (and probably won't before vacation).  Most likely guess would be that something about how `util/config/gather-clang-sysroot-arguments` is invoked or gets used to build up CFLAGS-like lists changed.  Any guesses?

(I specifically see the problem when hwloc tries to configure itself, but I'm not sure whether it's anything specific to hwloc, or whether that's just the first attempt to link with clang that we get during the build).

Tagging @ronawho on this, as he was offering ideas to debug this today.  I think this slipped by me until now simply because hwloc doesn't need to be rebuilt very often (?).

**Comment by mppf** [comment]
I don't know what's going wrong and would want to investigate further on a system showing the issue but I don't have one of those...

**Comment by bradcray** [comment]
In case others come across the last few comments, I've taken the conversation to https://github.com/Cray/chapel-private/issues/2069#issuecomment-894555342.

**Pull request merged**